File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        pre=1
        post=1
        res=[1 for i in range(len(nums))]
        for i in range(len(nums)):
            res[i]=pre
            pre*=nums[i]
        for i in range(len(nums)):
            res[len(nums)-1-i]*=post
            post*=nums[len(nums)-1-i]
        return res

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] ans = new int[n];
        int pre=1;
        // for pre
        for(int i=0;i<n;i++){
            ans[i] = pre;
            pre*=nums[i];
        }
        //  for post
        int post=1;
        for(int i=n-1;i>=0;i--){
            ans[i]*=post;
            post*=nums[i];
        }
    return ans;
    }
}